File: src/main/java/com/github/t1/yaml/model/Sequence.java
package com.github.t1.yaml.model;

import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;

@EqualsAndHashCode(callSuper = true)
public @Data class Sequence extends Collection {
    public static @Data class Item {
        boolean nl;
        Node node;
    }

    @Override public Sequence style(Style style) { super.style(style); return this; }

    private final List<Item> items = new ArrayList<>();

    public Sequence item(Node node) { return item(new Item().node(node)); }

    public Sequence item(Item item) {
        items.add(item);
        return this;
    }

    @Override public void guide(Visitor visitor) {
        visitor.visit(this);
        for (Item item : items) {
            visitor.enterSequenceItem(this, item);
            item.node.guide(visitor);
            visitor.leaveSequenceItem(this, item);
        }
        visitor.leave(this);
    }

    @Override public void canonicalize() {
        for (Item item : items)
            item.node.canonicalize();
    }

    public Item firstItem() { return items.get(0); }

    public Item lastItem() { return items.get(items.size() - 1); }
}

File: src/main/java/com/github/t1/yaml/model/Sequence.kt
package com.github.t1.yaml.model

data class Sequence(
    override var style: Style = Style.BLOCK,
    val items: MutableList<Item> = mutableListOf()
) : Collection(style) {

    data class Item(
        var node: Node,
        var nl: Boolean = false
    )

    fun item(node: Node): Sequence = item(Item(node))

    fun item(item: Item): Sequence {
        items.add(item)
        return this
    }

    fun firstItem(): Item = items[0]

    fun lastItem(): Item = items[items.size - 1]

    override fun guide(visitor: Visitor) {
        visitor.visit(this)
        for (item in items) {
            visitor.enterSequenceItem(this, item)
            item.node.guide(visitor)
            visitor.leaveSequenceItem(this, item)
        }
        visitor.leave(this)
    }
}
